src/coord/src/coord.rs

```diff
@@ -834,29 +844,67 @@ impl Coordinator {
                             );
                         }
                         if index_state.since != compaction_frontier {
+                            // Mark previous since for eviction.
+                            let mut updates: Vec<_> = index_state
+                                .since
+                                .elements()
+                                .iter()
+                                .map(|time| (*time, -1))
+                                .collect();
                             index_state.advance_since(&compaction_frontier);
-                            self.since_updates
-                                .push((name.clone(), index_state.since.clone()));
+                            // Mark new since for insertion.
+                            for time in index_state.since.elements() {
+                                updates.push((*time, 1));
+                            }
+                            // We always want `compaction_frontiers` to have exactly one timestamp tracking
+                            // the index's since, so send both the eviction and insertion in the same
+                            // update batch.
+                            self.update_compaction_frontier(*name, updates);
                         }
                     }
                 }
             }
         }
     }
 
+    /// Updates the tracked compaction frontier for an index.
+    ///
+    /// If the antichain progressed, it is added to the pending compactions.
+    fn update_compaction_frontier<I>(&mut self, name: GlobalId, updates: I)
+    where
+        I: IntoIterator<Item = (Timestamp, i64)>,
+    {
+        // Indexes are the only things that can be compacted.
+        assert!(self.catalog.get_by_id(&name).item_type() == CatalogItemType::Index);
```

Comment:
I am getting a panic here with a 2-thread workload that executes DDL and DML statements.

```
materialized v0.7.1-dev (d6525bd2a) listening on 0.0.0.0:6875...
2021-02-26T09:09:48.760945Z ERROR panic: no entry found for key
thread: <unnamed>
location: src/coord/src/catalog.rs:949:10
version: 0.7.1-dev (d6525bd2af5431801f6e2ff1d324e4b06fb72ba8)
backtrace:
   0: materialized::handle_panic
             at src/materialized/src/bin/materialized/main.rs:618:21
   1: core::ops::function::Fn::call
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:597:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:499:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/sys_common/backtrace.rs:141:18
   5: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   6: core::panicking::panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:92:14
   7: core::option::expect_failed
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/option.rs:1260:5
   8: core::option::Option<T>::expect
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/option.rs:349:21
      <alloc::collections::btree::map::BTreeMap<K,V> as core::ops::index::Index<&Q>>::index
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/alloc/src/collections/btree/map.rs:1957:9
      coord::catalog::Catalog::get_by_id
             at src/coord/src/catalog.rs:949:10
   9: coord::coord::Coordinator::update_compaction_frontier
             at src/coord/src/coord.rs:878:17
  10: coord::coord::Coordinator::sequence_end_transaction::{{closure}}
             at src/coord/src/coord.rs:2313:37
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/future/mod.rs:80:19
  11: coord::coord::Coordinator::sequence_plan::{{closure}}
             at src/coord/src/coord.rs:1646:21
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/future/mod.rs:80:19
      coord::coord::Coordinator::message_statement_ready::{{closure}}
             at src/coord/src/coord.rs:517:32
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/future/mod.rs:80:19
      coord::coord::Coordinator::serve::{{closure}}
             at src/coord/src/coord.rs:446:51
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/future/mod.rs:80:19
      tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.2.0/src/park/thread.rs:263:54
      tokio::coop::with_budget::{{closure}}
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.2.0/src/coop.rs:121:9
      std::thread::local::LocalKey<T>::try_with
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/thread/local.rs:272:16
      std::thread::local::LocalKey<T>::with
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/thread/local.rs:248:9
      tokio::coop::with_budget
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.2.0/src/coop.rs:114:5
      tokio::coop::budget
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.2.0/src/coop.rs:98:5
      tokio::park::thread::CachedParkThread::block_on
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.2.0/src/park/thread.rs:263:31
      tokio::runtime::enter::Enter::block_on
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.2.0/src/runtime/enter.rs:151:13
  14: tokio::runtime::thread_pool::ThreadPool::block_on
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.2.0/src/runtime/thread_pool/mod.rs:71:9
  15: tokio::runtime::Runtime::block_on
             at /home/pstoev/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.2.0/src/runtime/mod.rs:452:43
  16: coord::coord::serve::{{closure}}::{{closure}}
             at src/coord/src/coord.rs:3568:17
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/sys_common/backtrace.rs:125:18
  17: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/thread/mod.rs:474:17
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panic.rs:322:9
      std::panicking::try::do_call
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:381:40
      std::panicking::try
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:345:19
      std::panic::catch_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panic.rs:396:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/thread/mod.rs:473:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/ops/function.rs:227:5
  18: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/alloc/src/boxed.rs:1307:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/alloc/src/boxed.rs:1307:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/sys/unix/thread.rs:71:17
  19: start_thread
  20: clone 
```

Maybe the view is being yanked from underneath the transaction?

The issue reproduces easily for me in the context of the RQG framework. If you would like to run it, I can provide a machine that has it all set up or I can walk you through the installation.